tso-server/src/main/java/com/yahoo/omid/tso/PersistenceProcessorImpl.java

```diff
@@ -15,345 +15,216 @@
-                = new TimeoutBlockingWaitStrategy(config.getBatchPersistTimeoutMS(), TimeUnit.MILLISECONDS);
+        batchIDCnt = 0L;
 
-        persistRing = RingBuffer.createSingleProducer(
-                PersistEvent.EVENT_FACTORY, 1 << 20, timeoutStrategy); // 2^20 entries in ringbuffer
-        SequenceBarrier persistSequenceBarrier = persistRing.newBarrier();
-        BatchEventProcessor<PersistEvent> persistProcessor = new BatchEventProcessor<>(
-                persistRing,
-                persistSequenceBarrier,
-                this);
-        persistRing.addGatingSequences(persistProcessor.getSequence());
-        persistProcessor.setExceptionHandler(new FatalExceptionHandler(panicker));
+        batchPool = new BatchPool(config, batch);
 
-        ExecutorService persistExec = Executors.newSingleThreadExecutor(
-                new ThreadFactoryBuilder().setNameFormat("persist-%d").build());
-        persistExec.submit(persistProcessor);
+        this.batch = batchPool.getNextEmptyBatch();
 
-    }
+        timeoutMeter = metrics.meter(name("tso", "persist", "timeout"));
 
-    @Override
-    public void onEvent(PersistEvent event, long sequence, boolean endOfBatch) throws Exception {
-
-        switch (event.getType()) {
-            case COMMIT:
-                event.getMonCtx().timerStart("commitPersistProcessor");
-                // TODO: What happens when the IOException is thrown?
-                writer.addCommittedTransaction(event.getStartTimestamp(), event.getCommitTimestamp());
-                batch.addCommit(event.getStartTimestamp(), event.getCommitTimestamp(), event.getChannel(),
-                                event.getMonCtx());
-                break;
-            case ABORT:
-                sendAbortOrIdentifyFalsePositive(event.getStartTimestamp(), event.isRetry(), event.getChannel(),
-                                                 event.getMonCtx());
-                break;
-            case TIMESTAMP:
-                event.getMonCtx().timerStart("timestampPersistProcessor");
-                batch.addTimestamp(event.getStartTimestamp(), event.getChannel(), event.getMonCtx());
-                break;
-        }
+        persistRing = RingBuffer.<PersistBatchEvent>createSingleProducer(
+                PersistBatchEvent.EVENT_FACTORY, 1 << 20, new BusySpinWaitStrategy());
+        SequenceBarrier persistSequenceBarrier = persistRing.newBarrier();
 
-        if (batch.isFull() || endOfBatch) {
-            maybeFlushBatch();
+        numHandlers = config.getPersistHandlerNum();
+        handlers = new PersistenceProcessorHandler[numHandlers];
+        for (int i = 0; i < config.getPersistHandlerNum(); i++) {
+            handlers[i] = new PersistenceProcessorHandler(metrics, tsoHostAndPort, leaseManager, commitTable, reply, retryProc, panicker, config);
         }
 
-    }
+        WorkerPool<PersistBatchEvent> requestProcessor = new WorkerPool<PersistBatchEvent>(persistRing, persistSequenceBarrier,
```

Comment:
It might sounds like a minor naming pet peeve, but since we also have other request processors perhaps a better name would have been persistRequestProcessor or something along the lines